### PR TITLE
MDEV-31414: Implement optional lengths for string types

### PR DESCRIPTION
**Pull request by OsamaNabih**
Jira: https://jira.mariadb.org/browse/MDEV-31414
Adding support for SQL standard 2023's T081: `Optional string types maximum length`

This PR allows the user to not specify a length for a `VARCHAR` field. We default to making the field a `TEXT` type.

### CODE REVIEW

**Comment by FooBarrior** [comment]
Hi,
> Is 16383 reported as the maximum due to collation choice?

That is a nice guess! 

Our documentation [claims](https://mariadb.com/docs/server/reference/data-types/string-data-types/varchar#description) it's 21,844 characters, though I think it's a bit outdated, since we support 4-byte encodings for some time. So yes, it's got to be UINT16_MAX/4 - header_length, where header_length is something like 3.

File: sql/sql_type.cc [hunk 1]
```diff
@@ -2862,8 +2862,14 @@ Type_handler_varchar::Column_definition_set_attributes(
...
     break;
   case COLUMN_DEFINITION_ROUTINE_LOCAL:
-  case COLUMN_DEFINITION_TABLE_FIELD:
     break;
+  case COLUMN_DEFINITION_TABLE_FIELD:
+    /*
+      Support SQL Standard T081: "Optional string types maximum length"
+      Allows users to specify VARCHAR fields without a length
+    */
+    def->length= 16383;
```
Comment (FooBarrior) [inline sql/sql_type.cc]: Declare a constant in the appropriate place, like MAX_VARCHAR_SIZE, or maybe better MAX_VARCHAR_ESTIMATED_SIZE, signifying the imprecise nature.

I tried to look for the existing one and turns out it's pretty widely used in a raw literal form in the `sql` directory. Let's be the first good guys who'll add it there.

There is a constant in innodb -- we can't refer to it from `sql`, because this will reverse the dependencies. (innodb depends from sql, not vice-versa)

Comment (FooBarrior) [inline sql/sql_type.cc]: I'm pretty surprised this change would work on its own. Fine by me, if it does, though. Please add the test with a new syntax.

**Review by FooBarrior — CHANGES_REQUESTED** [review]
We follow the TDD paradigm - every feature should be covered with tests (and every bugfix, if possible). Include both .test and .result files in the commit.

I'm fine with a hard constant of 16383 in scope of this small work, we can always extend it and make a function of charset any time we want, since 16383 is expected to be the minimum of the maximums.

**Comment by FooBarrior** [comment]
You can try to make tha max value variable if you want. Though it may require some research. 

The problem is that we have (at least) three layers of charset settings:
* system-wide
* per-table
* per-field

One overrides another. At some point, Filed would get a deduced value of charset, and perhaps Create_field would get it, too. Though it's hard to say, when exactly. `Create_field` has `charset` field in its parent `Column_definition_attributes`, you may check if it's set to the right encoding at different points in time.

You'd have to test all three cases though:
* When a system-wide encoding is set
* When it's overriden with table
* When the latter is overriden with field charset.

Again, I'm NOT putting this as a requirement for this submission!

File: sql/sql_type.cc (hunk 1 above)
Comment (OsamaNabih) [reply sql/sql_type.cc]: Yes I expected to have it defined somewhere as it's used in several places, including the error that was reported, but I didn't find any so I hardcoded it for now.
My thought process was to get early feedback to verify my approach.
I now understand the better way of doing this is to start with the test.

I'll add a definition for `MAX_VARCHAR_ESTIMATED_SIZE`.
We can later investigate other places where the value `16383` is hardcoded and reference the same variable in a future PR. What do you think?

Comment (OsamaNabih) [reply sql/sql_type.cc]: As I am still familiarizing myself with the testing framework, I manually tested it by creating a table with a `VARCHAR` field without any parameters and creation was successful and running `DESCRIBE` returned `16383` as the result. This was just a sanity check for me locally.

I will add the `.test` and `.results` files. Is that sufficient to verify the code is behaving as expected?

**Comment by OsamaNabih** [comment]
> You can try to make tha max value variable if you want. Though it may require some research.
> 
> The problem is that we have (at least) three layers of charset settings:
> 
>     * system-wide
> 
>     * per-table
> 
>     * per-field
> 
> 
> One overrides another. At some point, Filed would get a deduced value of charset, and perhaps Create_field would get it, too. Though it's hard to say, when exactly. `Create_field` has `charset` field in its parent `Column_definition_attributes`, you may check if it's set to the right encoding at different points in time.
> 
> You'd have to test all three cases though:
> 
>     * When a system-wide encoding is set
> 
>     * When it's overriden with table
> 
>     * When the latter is overriden with field charset.
> 
> 
> Again, I'm NOT putting this as a requirement for this submission!

That is very informative.
I was mainly focusing on the `CREATE TABLE` scenario so I only paid attention to the `system-wide` encoding, but your comment made me realize there are other scenarios where `T081` applies as well.
For example `ALTER`, `CAST`, `CREATE TYPE`...etc 
For some of these scenarios, we need to handle the situation where the table/field were already created with a different `charset` than the `system-wide` one.

 To verify my understanding, we are okay with limiting the scope of  this PR to the `CREATE TABLE` scenario?

File: sql/sql_type.cc (hunk 1 above)
Comment (OsamaNabih) [reply sql/sql_type.cc]: > I'm fine with a hard constant of 16383 in scope of this small work, we can always extend it and make a function of charset any time we want, since 16383 is expected to be the minimum of the maximums.

I noticed this comment in `sql_type.cc` line `2796` which is concerning
>   In MariaDB the limit for VARCHAR is 65535 bytes.
  We could translate VARCHAR with no length to VARCHAR(65535), but
  it would mean that for multi-byte character sets we'd have to translate
  VARCHAR to MEDIUMTEXT, to guarantee 65535 characters.

 > Also we could translate VARCHAR to VARCHAR(16383), where 16383 is
  the maximum possible length in characters in case of mbmaxlen=4
  (e.g. utf32, utf16, utf8mb4). However, we'll have character sets with
  mbmaxlen=5 soon (e.g. gb18030). 
  
This was committed by `Alexander Barkov` 6 years ago, not sure if it is still relevant.
In the case of `gb18030`, the value of `16383` might not be the minimum after all. Is `gb18030` currently supported?

File: mysql-test/main/varchar_no_length.cnf [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Use default conf to set collation and charset
+!include include/default_my.cnf
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.cnf]: seems like a junk config

File: mysql-test/main/varchar_no_length.test [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+--echo # Test VARCHAR without length with various character sets
+--echo #
+
+CREATE TABLE t1 (a VARCHAR CHARSET utf8mb4);
+SHOW CREATE TABLE t1;
+DROP TABLE t1;
+
+CREATE TABLE t1 (a VARCHAR CHARSET latin1);
+SHOW CREATE TABLE t1;
+DROP TABLE t1;
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: Our conding standard (Apparently not mentioned in [CODING_STANDARDS](https://github.com/MariaDB/server/blob/main/CODING_STANDARDS.md) ) requires EOL before EOF.

Comment (FooBarrior) [reply mysql-test/main/varchar_no_length.test]: Add tests for:
CREATE TABLE ... CHARSET
CREATE TABLE ... ENGINE=InnoDB

same for ALTER TABLE

File: mysql-test/main/varchar_no_length.test [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# This tests the feature that allows VARCHAR columns to be created without
+# specifying a length, defaulting to the maximum VARCHAR length (65535)
+#
+
+--echo #
+--echo # SQL Standard T081: VARCHAR without explicit length specification
+--echo #
+
+--echo #
+--echo # Basic test: CREATE TABLE with VARCHAR without length
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: No need to duplicate captions. Leave the comment above, add `--echo #`, mention mdev and jira title. Remove this comment.

File: sql/sql_type.cc [hunk 5]
```diff
@@ -2862,8 +2862,16 @@ Type_handler_varchar::Column_definition_set_attributes(
...
     break;
   case COLUMN_DEFINITION_ROUTINE_LOCAL:
-  case COLUMN_DEFINITION_TABLE_FIELD:
     break;
+  case COLUMN_DEFINITION_TABLE_FIELD:
+    /*
+      Support SQL Standard T081: "Optional string types maximum length"
+      Allows users to specify VARCHAR fields without a length
+    */
+    def->length = def->charset && def->charset->mbmaxlen ?
```
Comment (FooBarrior) [inline sql/sql_type.cc]: I think we won't be able to determine the length correctly here. I think you'll realize that as soon as you'll get to trying the tests I requested above.

File: sql/sql_type.cc [hunk 6]
```diff
@@ -2862,8 +2862,16 @@ Type_handler_varchar::Column_definition_set_attributes(
...
+                  MAX_FIELD_VARCHARLENGTH / def->charset->mbmaxlen :
+                  MAX_FIELD_VARCHARLENGTH / thd->db_charset->mbmaxlen;
+    return false;
```
Comment (FooBarrior) [inline sql/sql_type.cc]: This line is likely correct though.

**Review by FooBarrior — CHANGES_REQUESTED** [review]

File: mysql-test/main/varchar_no_length.cnf (hunk 2 above)
Comment (OsamaNabih) [reply mysql-test/main/varchar_no_length.cnf]: I only used it to set a default charset/collation to have consistency for the tests. 
If the default charset/collation changes in the future, this may cause this test to fail, even when the logic is still correct (maybe it's better to have it fail and the dev should verify anyway?).
Also, I wasn't sure if the server's choice for the default charset changes based on the machine settings (such as language), in that case the test my fail for other machines, so I thought having it set in the .opt is better.
What do you think?

File: sql/sql_type.cc (hunk 5 above)
Comment (OsamaNabih) [reply sql/sql_type.cc]: Good point. I am only considering column and DB charsets, I should also consider the table charset.

So the order is column > table > DB
Is it possible DB charset is not set? In that case I can consider session charset and system-wide charsets as fallback, would that be necessary?

File: mysql-test/main/varchar_no_length.test (hunk 3 above)
Comment (OsamaNabih) [reply mysql-test/main/varchar_no_length.test]: Thanks! I missed the Coding Standards document.
I will make sure to adhere to it for all files and git commits.

File: sql/sql_type.cc (hunk 1 above)
Comment (OsamaNabih) [reply sql/sql_type.cc]: `gb18030` is not yet supported, this concern is not currently valid.

File: mysql-test/main/varchar_no_length.cnf (hunk 2 above)
Comment (OsamaNabih) [reply mysql-test/main/varchar_no_length.cnf]: Removed.

File: sql/sql_type.cc (hunk 5 above)
Comment (OsamaNabih) [reply sql/sql_type.cc]: The division is not an issue, DB charset is set.

File: mysql-test/main/varchar_no_length.result [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+# SQL Standard T081: VARCHAR without explicit length specification
+#
+#
+# Basic test: CREATE TABLE with VARCHAR without length or charset (scenario 1a)
+#
+CREATE TABLE t1 (a VARCHAR);
+SHOW CREATE TABLE t1;
+Table	Create Table
+t1	CREATE TABLE `t1` (
+  `a` text DEFAULT NULL
```
Comment (gkodinov) [inline mysql-test/main/varchar_no_length.result]: my own 2 cents here: it would look odd to people saying "a VARCHAR" and getting "a TEXT" back as a result of SHOW CREATE.

I'd recognize this in SHOW CREATE and print back VARCHAR.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thanks for working on this. This is a preliminary review. 

Please fix some sanity checks: 

* Make sure this is a single commit and it has a proper commit message, according to CODING_STANDARDS.md.
* Make sure there's no failing tests as a result of your change. I can see a couple in the buildbot results.

File: mysql-test/main/varchar_no_length.result (hunk 7 above)
Comment (FooBarrior) [reply mysql-test/main/varchar_no_length.result]: We've been discussing these approaches with @vuvova - making it substituted in show create means saving this information in frm somehow, which seems an overkill for such a slight improvement.

That is, we first agreed that this should be an auto-calculated varchar with a max possible size. But then, turns out we have to mind the field encoding to calculate the max possible field size, and moreover it won't be possible to add more than one such column.

So Serg said, ok let it be TEXT under the hood.

I haven't formed any strong opinion against being it text. It confuses me with imperfection, but other than that I'm fine.

Comment (gkodinov) [reply mysql-test/main/varchar_no_length.result]: I do not advocate one way or another. Just making some usability remarks. 
But, no matter what we settle on, we at least need a very good documentation (preferably in the jira itself) on what gets stored where in what case. 
I.e. what length goes into the .frm for "a VARCHAR"? 
Note that this seems to be allowed syntax by the parser even today. I can e.g. write:

```
MariaDB [test]> create table t1 (a char);
Query OK, 0 rows affected (0.035 sec)

MariaDB [test]> show create table t1\G
*************************** 1. row ***************************
       Table: t1
Create Table: CREATE TABLE `t1` (
  `a` char(1) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_uca1400_ai_ci

```

Comment (vuvova) [reply mysql-test/main/varchar_no_length.result]: I read in https://mariadb.com/docs/server/reference/sql-statements/data-definition/create/silent-column-changes

> If strict SQL mode is not enabled (by default, it is), a VARCHAR column longer than 65535 become TEXT, and a VARBINARY columns longer than 65535 becomes a BLOB.

That's why I though as it's already converts VARCHAR→TEXT, let's use that. At least it'll be consistent.

Comment (gkodinov) [reply mysql-test/main/varchar_no_length.result]: Serg, 

So you're thinking that:

`CREATE TABLE t1 (a VARCHAR) `

should be equivalent to 

`CREATE TABLE t1 (a TEXT)`

?

As in, you treat a missing length as passing down a maximum length of some sort, right? 

Note that right now, without any change:

`CREATE TABLE t1 (a CHAR) `

is equivalent to 

`CREATE TABLE t1 (a CHAR(1))`

So the above change would not be consistent with that. And it won't be consistent with postgres too: according to the jira text, pgsql treats VARCHAR the same way as CHAR when length is missing.

This is why it seemed more logical to me to not go to TEXT. 

But, again, I'm just trying to think out-loud here. 

Most importantly: whatever you decide on, let's please make sure is well documented in the MDEV and into the commit itself.

Comment (vuvova) [reply mysql-test/main/varchar_no_length.result]: Strange. I was sure that behavior was somewhere in the standard and I took it from there. But now I'm seeing

> 5) If \<length> is omitted, then a \<length> of 1 (one) is implicit.

This is from Syntax Rules for 6.1 \<data type>, Part 2: Foundation, SQL16.

Does not seem to be "implementation defined", just 1 (one). Makes everything simper, doesn't it?

Comment (OsamaNabih) [reply mysql-test/main/varchar_no_length.result]: I recall reading that this is implementation dependent. It can be as low as 1, but I believe Postgres goes the maximum size way.

According to the docs here

https://www.postgresql.org/docs/current/datatype-character.html

> If character varying (or varchar) is used without length specifier, the type accepts strings of any length

For `char`, they default to `char(1)`.

> If character (or char) lacks a specifier, it is equivalent to character(1).

Not saying we should do it the same way.

File: mysql-test/main/varchar_no_length.test [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+SHOW CREATE TABLE t1;
+DROP TABLE t1;
+
+--echo
+--echo #
+--echo # Test VARCHAR COMPRESSED without length
+--echo #
+CREATE TABLE t1 (a VARCHAR COMPRESSED);
+SHOW CREATE TABLE t1;
+DROP TABLE t1;
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: tests should have a single empty line in the end of file

File: mysql-test/main/varchar_no_length.result [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+SHOW CREATE TABLE t1;
+Table	Create Table
+t1	CREATE TABLE `t1` (
+  `id` int(11) DEFAULT NULL,
+  `a` text DEFAULT NULL
+) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_uca1400_ai_ci
+DROP TABLE t1;
+
+#
+# Corner Case: VARBINARY without length should become BLOB
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.result]: This isn't a corner case. The sql standard rules spread to VARBINARY and NVARCHAR as well.
If it wouldn't then you'd have to fix it, but not accept side-effects as collateral damage

File: mysql-test/main/varchar_no_length.test [hunk 10]
```diff
@@ -0,0 +1,123 @@
+#
+# MDEV-31414
+# Test for SQL Standard T081: "Optional string types maximum length"
+# This tests the feature that allows VARCHAR columns to be created without
+# specifying a length, defaulting to the maximum VARCHAR length (65535)
+#
+# We have 3 scenarios for resolving charset and collation
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: I think this is not actual anymore, as we've moved to TEXT/BLOB substitution

File: mysql-test/main/varchar_no_length.test [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+# Check the comment section in "lex_charset.cc" at 
+# CHARSET_INFO *Lex_exact_charset_extended_collation_attrs_st::
+#                resolved_to_character_set
+
+--echo #
+--echo # SQL Standard T081: VARCHAR without explicit length specification
+--echo #
+
+--echo #
+--echo # Basic test: CREATE TABLE with VARCHAR without length or charset (scenario 1a)
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: Remove the "Basic test" and "Test" wordings hereafter.

File: mysql-test/main/varchar_no_length.test [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+--echo
+--echo #
+--echo # Test VARCHAR without length with specific column charsets or collations (scenario 2a)
+--echo #
+
+CREATE TABLE t1 (a VARCHAR CHARSET utf8mb4);
+SHOW CREATE TABLE t1;
+DROP TABLE t1;
+
+--echo
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: Forgot the hashes

File: mysql-test/main/varchar_no_length.test [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+--echo #
+--echo # Basic test: CREATE TABLE with charset (Scenario 3)
+--echo #
+CREATE TABLE t1 (a VARCHAR) CHARSET latin1;
+SHOW CREATE TABLE t1;
+DROP TABLE t1;
+
+--echo
+--echo #
+--echo # Test ALTER TABLE MODIFY COLUMN with VARCHAR without length
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: I think these comments tell a pretty obvious information. That makes the tests harder to read.

Remove the obvious comments or provide useful information in them

**Review by FooBarrior — CHANGES_REQUESTED** [review]
Add more details in the commit comments. Mention SQL standard, its rule about implementation-defined lengths and describe our approach.
Remove Signed-off-by.

File: mysql-test/main/varchar_no_length.test [hunk 14]
```diff
@@ -0,0 +1,116 @@
+#
+# MDEV-31414
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: --echo

**Review by FooBarrior — CHANGES_REQUESTED** [review]

File: mysql-test/main/varchar_no_length.test (hunk 10 above)
Comment (OsamaNabih) [reply mysql-test/main/varchar_no_length.test]: I understood this comment as referring to the part regarding using maximum length VARCHAR and removed it accordingly.
Maybe I have misunderstood.
Are you referring to the charset comments/tests as also no longer relevant and should be removed?

Comment (FooBarrior) [reply mysql-test/main/varchar_no_length.test]: I am referring to the three scenarios fact. We have three layers of charset derivation, yes, but how is it relevant to your current implementation?

File: mysql-test/main/varchar_no_length.test [hunk 15]
```diff
@@ -0,0 +1,116 @@
+--echo #
+--echo # MDEV-31414
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: --echo # MDEV-31414: Implement optional lengths for string types

**Review by FooBarrior — CHANGES_REQUESTED** [review]

File: mysql-test/main/varchar_no_length.test (hunk 11 above)
Comment (FooBarrior) [reply mysql-test/main/varchar_no_length.test]: This still isn't resolved.

File: mysql-test/main/varchar_no_length.test [hunk 16]
```diff
@@ -0,0 +1,116 @@
+--echo #
```
Comment (FooBarrior) [inline mysql-test/main/varchar_no_length.test]: rename the test to type_varchar_no_length

**Review by FooBarrior — CHANGES_REQUESTED** [review]
Please go through the test carefully, by yourself. Because I see that my edit requests stay unfulfilled, yet marked as resolved.

File: mysql-test/main/varchar_no_length.test (hunk 10 above)
Comment (OsamaNabih) [reply mysql-test/main/varchar_no_length.test]: It is just making sure existing behavior is intact. I can remove them to not bloat the test-suite, they are probably covered by other tests.

**Review by FooBarrior — APPROVED** [review]
LGTM

**Review by gkodinov — APPROVED** [review]
LGTM

**Comment by gkodinov** [comment]
Note that, before being merged, this needs to go through testing. So please, have some more patience.

**Review by gkodinov — CHANGES_REQUESTED** [review]
I've received the following request from the QA engineer: 

Please add tests for other places where the length-less type can be used: 

* local stored procedure/function/trigger variables, via DECLARE
* stored procedure/function parameters
* local  stored function return type
* JSON_TABLE
* dynamic columns
* virtual columns

Note that the SP test should be done in both Oracle and native mode.

**Comment by OsamaNabih** [comment]
> I've received the following request from the QA engineer: 
> 
> Please add tests for other places where the length-less type can be used: 
> 
> * local stored procedure/function/trigger variables, via DECLARE
> * stored procedure/function parameters
> * local  stored function return type
> * JSON_TABLE
> * dynamic columns
> * virtual columns
> 
> Note that the SP test should be done in both Oracle and native mode. 
> 

I'll start looking into those
Are these required as a part of the scope of this PR or for a future PR?

**Comment by gkodinov** [comment]
This is what is needed to:
* document the feature fully
* test it properly

So, I guess, for this PR. Note that it's about *documenting and test coverage* for the *current behavior*. It's not about changing it yet.

**Comment by gkodinov** [comment]
Changing the state of this pull request to Draft due to the lack of reply to my questions. Please move it back to "open" when/if you intend to resume working on it.

**Review by gkodinov — APPROVED** [review]
LGTM. Stand by for testing.

**Comment by mariadb-RameshSivaraman** [comment]
Found the following issues during testing. Please take a look and fix them.
https://jira.mariadb.org/browse/MDEV-40049
https://jira.mariadb.org/browse/MDEV-40051
https://jira.mariadb.org/browse/MDEV-40079

**Comment by gkodinov** [comment]
FYI: According to our development cycle all testing of features must be completed before 30 Apr, 30 Jul, 30 Nov and 31 Jan (approximate release dates). You might get comments from the tester before that date or your feature will be approved as is and merged. But it'll all likely happen closer to the release date, not right now.
